In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from matplotlib_inline.backend_inline import set_matplotlib_formats
set_matplotlib_formats('svg')

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import cross_val_score
from sklearn.metrics import roc_curve, auc
df = pd.read_csv('/content/train.csv')
df.head()


Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

In [13]:
y_pred = model.predict(X_test)
display(y_pred)

array([113410.67255298, 305081.87775899, 135904.78562983, 205424.67564124,
       227502.68349004, 121157.48079629, 205577.98056584, 183787.20378269,
       121157.48079629, 147219.22233196, 185971.75622995,  93229.5689926 ,
       121965.26049018, 192005.28203666, 208129.35962356, 133824.25948958,
       211250.14883393, 146931.57417313, 134240.36471763, 186179.80884397,
       219884.33231595, 197726.72892233, 185867.72992294, 126334.3653847 ,
       181186.54610738, 183707.60823788, 171720.15216927, 118844.47127982,
       199183.09722051, 199623.63321076, 132551.51304323, 245162.72491993,
       383364.40832164, 127998.78629689, 202511.9390449 , 120484.46142981,
       200376.66521523, 206232.45533514, 241160.86372906, 120716.94480604,
       124645.5137103 , 251820.40856871, 101135.56832553, 244746.61969188,
       105088.56799199, 208912.70855526, 123837.7340164 , 101655.69986059,
       271169.30167299, 160846.66855057, 105088.56799199, 221989.28921839,
       137960.88100787, 3

# Task
Implement a linear regression model to predict house prices using 'GrLivArea', 'BedroomAbvGr', and 'FullBath' as features and 'SalePrice' as the target variable from the file "train.csv". Split the data into training and testing sets, train the model, and evaluate its performance.

## Feature selection

### Subtask:
Select the 'GrLivArea', 'BedroomAbvGr', and 'FullBath' columns as features and 'SalePrice' as the target variable.


**Reasoning**:
Select the specified feature and target columns from the dataframe.



In [5]:
features = ['GrLivArea', 'BedroomAbvGr', 'FullBath']
target = 'SalePrice'
X = df[features]
y = df[target]

## Data splitting

### Subtask:
Split the data into training and testing sets.


**Reasoning**:
Split the feature and target data into training and testing sets for model training and evaluation.



In [8]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

## Model training

### Subtask:
Train a linear regression model using the training data.


**Reasoning**:
Train a linear regression model using the training data.



In [10]:
model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression()

## Model evaluation

### Subtask:
Evaluate the model's performance on the testing data using appropriate metrics.


**Reasoning**:
Make predictions on the test set and calculate the MSE and R-squared metrics.



In [12]:
y_pred = model.predict(X_test)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)
print(f'Mean Squared Error: {mse}')
print(f'R-squared: {r2}')

Mean Squared Error: 2806426667.247853
R-squared: 0.6341189942328371


## Summary:

### Data Analysis Key Findings
* The linear regression model was trained using 'GrLivArea', 'BedroomAbvGr', and 'FullBath' as features and 'SalePrice' as the target variable.
* The data was split into training and testing sets with a test size of 20%.
* The Mean Squared Error (MSE) of the model on the testing data is approximately 2.81e9.
* The R-squared score of the model on the testing data is approximately 0.634.

### Insights or Next Steps
* The R-squared value of 0.634 indicates that approximately 63.4% of the variance in house prices can be explained by the selected features.
* Consider exploring additional features or different regression models to potentially improve the prediction performance and achieve a higher R-squared score.


# Task
Predict the 'SalePrice' using the trained linear regression model on the data from "test.csv".

## Load test data

### Subtask:
Load the data from "test.csv" into a dataframe.


**Reasoning**:
Load the data from "test.csv" into a pandas DataFrame and display the first few rows.



In [14]:
df_test = pd.read_csv('/content/test.csv')
display(df_test.head())

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0  1461          20       RH         80.0    11622   Pave   NaN      Reg   
1  1462          20       RL         81.0    14267   Pave   NaN      IR1   
2  1463          60       RL         74.0    13830   Pave   NaN      IR1   
3  1464          60       RL         78.0     9978   Pave   NaN      IR1   
4  1465         120       RL         43.0     5005   Pave   NaN      IR1   

  LandContour Utilities  ... ScreenPorch PoolArea PoolQC  Fence MiscFeature  \
0         Lvl    AllPub  ...         120        0    NaN  MnPrv         NaN   
1         Lvl    AllPub  ...           0        0    NaN    NaN        Gar2   
2         Lvl    AllPub  ...           0        0    NaN  MnPrv         NaN   
3         Lvl    AllPub  ...           0        0    NaN    NaN         NaN   
4         HLS    AllPub  ...         144        0    NaN    NaN         NaN   

  MiscVal MoSold  YrSold  SaleType  SaleCondition  
0       0      6    2010        WD         Normal  
1   12500      6    2010        WD         Normal  
2       0      3    2010        WD         Normal  
3       0      6    2010        WD         Normal  
4       0      1    2010        WD         Normal  

[5 rows x 80 columns]

## Feature selection for test data

### Subtask:
Select the same features ('GrLivArea', 'BedroomAbvGr', and 'FullBath') from the test data.


**Reasoning**:
Select the specified feature columns from the test dataframe.



In [15]:
features = ['GrLivArea', 'BedroomAbvGr', 'FullBath']
X_test_features = df_test[features]
display(X_test_features.head())

GrLivArea  BedroomAbvGr  FullBath
0        896             2         1
1       1329             3         1
2       1629             3         2
3       1604             3         2
4       1280             2         2

## Predict on test data

### Subtask:
Use the trained linear regression model to predict 'SalePrice' on the test data.


**Reasoning**:
Use the trained linear regression model to predict 'SalePrice' on the test data.



In [16]:
y_pred_test = model.predict(X_test_features)
display(y_pred_test)

array([122173.31310421, 140561.53868318, 201783.75489581, ...,
       102983.61108955, 103216.09446577, 240377.51479736])

## Summary:

### Data Analysis Key Findings

* The test data from "test.csv" was successfully loaded into a pandas DataFrame.
* The features 'GrLivArea', 'BedroomAbvGr', and 'FullBath' were successfully selected from the test data.
* The trained linear regression model was used to predict 'SalePrice' on the test data, resulting in an array of predicted values.

### Insights or Next Steps

* The predicted 'SalePrice' values can now be used for further analysis or submission to a competition.
* It would be beneficial to evaluate the performance of the model on the test data if the true 'SalePrice' values were available.


In [17]:
# Create a DataFrame with 'Id' and predicted 'SalePrice'
submission_df = pd.DataFrame({'Id': df_test['Id'], 'SalePrice': y_pred_test})

# Display the first few rows of the submission DataFrame
display(submission_df.head())

# Save the submission DataFrame to a CSV file
submission_df.to_csv('submission.csv', index=False)

Id      SalePrice
0  1461  122173.313104
1  1462  140561.538683
2  1463  201783.754896
3  1464  199183.097221
4  1465  192133.739106